### ISSUE TITLE
Support nodeSelector for Prometheus TPR

### ISSUE THREAD

**Issue by iterion** [coreos/prometheus-operator#107]
I'd like to control which nodes my Prometheus instance will be scheduled onto. But, it seems that `nodeSelector` is not available in the Prometheus TPR, which, I assume, would be the correct place to specify it.

**Comment by brancz** [comment]
This is definitely something we want to support. It should be pretty straight forward, so I'll try to get it into the next release.

**Comment by iterion** [comment]
Cool, thanks! I'd also love to contribute, so if you're busy with other stuff I'd certainly like to try to knock it out.

I assume we would modify the `PrometheusSpec` to look like the following:
```golang
// PrometheusSpec holds specification parameters of a Prometheus deployment.
type PrometheusSpec struct {
	ServiceMonitorSelector *metav1.LabelSelector   `json:"serviceMonitorSelector"`
	Version                string                  `json:"version"`
	Paused                 bool                    `json:"paused"`
	BaseImage              string                  `json:"baseImage"`
	Replicas               int32                   `json:"replicas"`
	Retention              string                  `json:"retention"`
	ExternalURL            string                  `json:"externalUrl"`
	Storage                *StorageSpec            `json:"storage"`
	Alerting               AlertingSpec            `json:"alerting"`
	Resources              v1.ResourceRequirements `json:"resources"`
	NodeSelector           map[string]string       `json:"nodeSelector,omitempty"
}
```

And then just copy the `NodeSelector` map into the `StatefulSet` that is generated?

**Comment by brancz** [comment]
You can most certainly give it a go! Yes indeed it would go into the `PodTemplate` of the `StatefulSet`. Looking forward to reviewing! 🙂 🎉

**Comment by fabxc** [comment]
My only concern at this point is, that there is also node affinity and anti-affinity: https://kubernetes.io/docs/user-guide/node-selection/

Ideally we wouldn't carry two concepts where one is a superset of the other. Do we want both, or just the newer and more powerful one?

They are still alpha however and still done via JSON encoded annotations. For storage volumes we had a similar case, where we simply went ahead and implemented proper YAML fields for that. That has the issue that we have to break those, if the annotation format changes.

We could also carry it as an annotation to the TPR as well for the time being and propagate it down. This ties into #95 then.

**Comment by iterion** [comment]
Took a stab at it with #109.

@fabxc It seems to me that supporting both would be important. For my use case nodeSelector is sufficiently powerful. In addition, `nodeSelector` is already part of the v1 spec. So, it's not going anywhere at least until v2. That said, passing along annotations from the prometheus TPR would support a large number of alpha features so it would also be beneficial.

**Comment by brancz** [comment]
Looking at the documentation 

> If you specify both nodeSelector and nodeAffinity, both must be satisfied for the pod to be scheduled onto a candidate node.

I think it should be safe to start with a `nodeSelector`, and extend it with `nodeAffinity` once there is a request for it, in the hope that by then it will have been promoted from alpha stage.

**Comment by brancz** [comment]
Yes we're gonna stay away from node affinity until 1.6 when it will be promoted to beta and become an object field ... https://groups.google.com/forum/m/#!msg/kubernetes-dev/-fwfHQtm9GE/H1TU4yqiBAAJ

**Comment by fabxc** [comment]
Makes all sense. Just would be a bit unnecessary if nodeSelectors were just
kept around due to 1.x in k8s and effectively superseded.
But fine with me to add them now. If it really bothers us, we could always
to some auto-migration later – or just keep the few lines of code and docs
around; I might be too concerned with minimalism.

**Comment by brancz** [comment]
Closed by #109 and #121.
